python/metric_pipeline_utils/stellar_locus.py

```diff
@@ -0,0 +1,214 @@
+import scipy.stats as scipyStats
+import scipy.odr as scipyOdr
+from lsst.pipe.base import Struct
+from metric_pipeline_utils.filtermatches import filterMatches
+from lsst.afw.table import GroupView
+import astropy.units as u
+
+
+def stellarLocusResid(gmags, rmags, imags, **filterargs):
+
+    gr = gmags-rmags
+    ri = rmags-imags
+
+    # Also trim large values of r-i, since those will skew the linear regression
+    okfitcolors = (gr < 1.1) & (gr > 0.3) & (np.abs(ri) < 1.0) & np.isfinite(gmags) & np.isfinite(rmags) & np.isfinite(imags)
+    # Eventually switch to using orthogonal regression instead of linear (as in pipe-analysis)?
+
+    slope, intercept, r_value, p_value, std_err = scipyStats.linregress(gr[okfitcolors],
+                                                                        ri[okfitcolors])
+    p2p1coeffs = p2p1CoeffsFromLinearFit(slope, intercept, 0.3, slope*0.3+intercept)
+    p1coeffs = p2p1coeffs.p1Coeffs.copy()
+    # hack to put the zeros in for u, z coeffs
+    p1coeffs.insert(0, 0.0)
+    p1coeffs.insert(4, 0.0)
+    p2coeffs = list(p2p1coeffs.p2Coeffs.copy())
+    p2coeffs.insert(0, 0.0)
+    p2coeffs.insert(4, 0.0)
+    umags = np.zeros(len(gmags))
+    zmags = np.zeros(len(gmags))
+    p1_fit = calcP1P2([umags, gmags, rmags, imags, zmags], p1coeffs)
+    p2_fit = calcP1P2([umags, gmags, rmags, imags, zmags], p2coeffs)
+    okp1_fit = (p1_fit < 0.6) & (p1_fit > -0.2)
+
+    # Do a second iteration, removing large (>3 sigma) outliers in p2:
+    clippedStats = calcQuartileClippedStats(p2_fit[okp1_fit], 3.0) 
+    keep = (np.abs(p2_fit) < clippedStats.clipValue)
+
+    slope, intercept, r_value, p_value, std_err = scipyStats.linregress(gr[okfitcolors & keep],
+                                                                        ri[okfitcolors & keep])
+    p2p1coeffs = p2p1CoeffsFromLinearFit(slope, intercept, 0.3, slope*0.3+intercept)
+    p1coeffs = p2p1coeffs.p1Coeffs.copy()
+    # hack to put the zeros in for u, z coeffs
+    p1coeffs.insert(0, 0.0)
+    p1coeffs.insert(4, 0.0)
+    p2coeffs = list(p2p1coeffs.p2Coeffs.copy())
+    p2coeffs.insert(0, 0.0)
+    p2coeffs.insert(4, 0.0)
+    p1_fit = calcP1P2([umags, gmags, rmags, imags, zmags], p1coeffs)
+    p2_fit = calcP1P2([umags, gmags, rmags, imags, zmags], p2coeffs)
+    okp1_fit = (p1_fit < 0.6) & (p1_fit > -0.2)
+
+    return p1_fit[okp1_fit], p2_fit[okp1_fit], p1coeffs, p2coeffs
+
+
+def calcP1P2(mags, coeffs):
+    # P1 =A′u+B′g+C′r+D′i+E′z+F′
+    # P2′=Au+Bg+Cr+Di+Ez+F
+    # umag = np.array(mags['umag'],dtype='float')
+    # gmag = np.array(mags['gmag'],dtype='float')
+    # rmag = np.array(mags['rmag'],dtype='float')
+    # imag = np.array(mags['imag'],dtype='float')
+    # zmag = np.array(mags['zmag'],dtype='float')
+    p1p2 = float(coeffs[0])*mags[0] + float(coeffs[1])*mags[1] +\
+        float(coeffs[2])*mags[2] + float(coeffs[3])*mags[3] +\
+        float(coeffs[4])*mags[4] + float(coeffs[5])
+    return p1p2
+
+
+def getCoeffs():
```

Comment:
This function does not seem to be called anywhere (?)

Reply:
It isn't currently called anywhere, but I wanted to have it here so that I can add a fallback to these coefficients if the linear fit to the stellar locus fails. I'll do that later.